yt_dlp/extractor/niconico.py

```diff
@@ -411,17 +424,97 @@ def _real_extract(self, url):
                     raise
                 raise ExtractorError(re.sub(r'\s+', ' ', error_msg), expected=True)
 
+        club_joined = traverse_obj(api_data, ('channel', 'viewer', 'follow', 'isFollowed', {bool}))
+        if club_joined is None:
+            fail_msg = self._html_search_regex(
+                r'<p[^>]+\bclass="fail-message"[^>]*>(?P<msg>.+)</p>',
+                webpage, 'fail message', default=None, group='msg')
+            if fail_msg:
+                self.raise_login_required(re.sub(r'\s+', ' ', fail_msg), metadata_available=True)
+        elif not club_joined:
+            self.raise_login_required('This video is for members only', metadata_available=True)
+
         formats = []
 
         def get_video_info(*items, get_first=True, **kwargs):
             return traverse_obj(api_data, ('video', *items), get_all=not get_first, **kwargs)
 
-        quality_info = api_data['media']['delivery']['movie']
-        session_api_data = quality_info['session']
-        for (audio_quality, video_quality, protocol) in itertools.product(quality_info['audios'], quality_info['videos'], session_api_data['protocols']):
-            fmt = self._extract_format_for_quality(video_id, audio_quality, video_quality, protocol)
-            if fmt:
-                formats.append(fmt)
+        dmc_data = traverse_obj(api_data, (
+            'media', 'delivery', 'movie', {
+                'audios': 'audios',
+                'videos': 'videos',
+                'protocols': ('session', 'protocols'),
+            }, {lambda data: data if all(key in data for key in ['audios', 'videos', 'protocols']) else None},
+        ))
+        if dmc_data:
+            for (audio_quality, video_quality, protocol) in itertools.product(
+                    dmc_data['audios'], dmc_data['videos'], dmc_data['protocols']):
+                fmt = self._extract_format_for_quality(video_id, audio_quality, video_quality, protocol)
+                if fmt:
+                    formats.append(fmt)
+
+        dms_data = traverse_obj(api_data, ({
+            'videos': ('media', 'domand', 'videos', ..., {
+                lambda item: item['id'] if item['isAvailable'] else None
+            }),
+            'audios': ('media', 'domand', 'audios', ..., {
+                lambda item: (item['id'], {
+                    'format_id': str(parse_bitrate(item['id'])),
+                    'abr': float_or_none(item['bitRate'], scale=1000),
+                    'asr': item['samplingRate'],
+                    'acodec': 'aac',
+                    'ext': 'm4a',
+                }) if item['isAvailable'] else None
```

Comment:
I think once you have to use a transformation to construct an entire format_dict then it's time to break this up and use separate variables/traversals.

Also, can you show an example of what the POST data that we send with the API request should look like? Maybe I'm misreading the code, but it looks like we are sending a tuple of the `id` and the format dict, and that seems wrong

Reply:
> Can you show an example of what the POST data that we send with the API request should look like?

Of course. The POST request is:

```bash
curl -fsSL 'https://nvapi.nicovideo.jp/v1/watch/sm.../access-rights/hls?actionTrackId=...' \
  -H 'x-access-right-key: ...' \
  -H 'x-frontend-id: 6' \
  -H 'x-frontend-version: 0' \
  -H 'x-request-with: https://www.nicovideo.jp' \
  --data '{"outputs": [
    ["video-h264-1080p","audio-aac-128kbps"],
    ["video-h264-720p","audio-aac-128kbps"],
    ["video-h264-480p","audio-aac-128kbps"],
    ["video-h264-360p","audio-aac-128kbps"],
    ["video-h264-144p","audio-aac-128kbps"]
  ]}' | jq .
```

The response is:

```JSON
{
  "meta": {
    "status": 201
  },
  "data": {
    "contentUrl": "https://delivery.domand.nicovideo.jp/hlsbid/.../playlists/variants/....m3u8?...",
    "createTime": "...",
    "expireTime": "..."
  }
}

```

<br>

> Maybe I'm misreading the code, but it looks like we are sending a tuple of the `id` and the format dict, and that seems wrong

Well, I'm only sending all IDs, not the format dicts.

<br>

I've split the extraction, please re-review it if you have free time.